[WIP] Bulk scene import

WithoutPants opened the pull request:
Built off Infinite's branch and based on the design in #116.

- Added new user roles for submitting and finalising imports.
- Limited to a single pending import per user.
- Supports csv only
- Added import top-level menu that is only visible for authorised users

![image](https://user-images.githubusercontent.com/53250216/130216219-006a27ca-0652-4bbd-b907-904bc6473c80.png)

Allows configuration of scene field to input data file field or a fixed value (ie what "Main Studio" was apparently doing). Allows multiple per-field regex replacements, which are performed in order. Configurations can be saved and loaded to/from a json file.

Not pictured is the list delimiter field. For fields that allow multiple values (tags, performers), this string is used to delimit values. This is applied after the regex replacements.

After submitting (or if there is a pending import):

![image](https://user-images.githubusercontent.com/53250216/130216652-47d3201d-e840-4c34-b02e-a68c14821203.png)

Allows browsing of the data to be imported. Not shown in this image, but where a studio/tag/performer is matched, the text is normal white and an applicable icon is shown. The scene images are also not shown in the above image, but are supported.

Pending imports can be completed or aborted. 

TODO:
- allowing changing of performer/tag/studio name to object mappings isn't added yet. There could potentially be a large number of these to deal with, so need to think about an appropriate way to deal with it
- importing scenes with an image URL is super slow and probably not done in a sustainable way - it currently just slurps all the URLs as needed. This probably needs to be offloaded to a separate task and throttled appropriately. An alternative is to accept encoded images in the data files and just use those instead of scraping them from the source.

Leopere committed:
Initial commit

Leopere committed:
Update README.md

Leopere committed:
Create empty compiler Dockerfile for Dockerhub

WithoutPants committed:
Initial performer prototype (#1)

* Initial performer prototype

* Drop file: from migration url to fix panic

* Remove vendor requirement

Leopere committed:
Create FUNDING.yml

For consistency adding the same FUNDING.yml from the main repo.

WithoutPants committed:
Add studio, tags, scene (#3)

* Add tags
* Add studio
* Add scenes support
* Add integration test
* Add gitattributes. Add test targets
* Add DB interface and refactor
* Add performer image
* Replace checksums with fingerprints
* Update dependencies
* Make performers unique on name/disambiguation
* Add first draft of README

WithoutPants committed:
Add support for postgres database (#4)

* Generalise database
* Add postgres migration
* Add optional support for postgres database
* Add postgres documentation to README

Leopere committed:
Update issue templates (#2)

* Update issue templates

* Added censorship clause to template.

Commit:
Change type of IDs from int64 to uuid  (#7)

* Change type of IDs from int64 to uuid
* Rename insertJoins -> insertObjects

WithoutPants committed:
Implement MultiIDCriterion filters for scene (#8)

Commit:
Align backend and frontend (#10)

* Update URLs

Commit:
Frontend for stash-box (#5)

WithoutPants committed:
Fix integration tests (#12)

* Fix test compile error and panic

* Remove irrelevant tests

* Fix NULLS LAST change

* Support integration test against postgres

* Fix postgres test failure

WithoutPants committed:
Add User management (#13)

* Add user
* Generate root user
* Add create/update validation
* Disallow removing root user
* Fix packr deprecation message
* Change authorisation to user apikey
* Return error message when apikey fails
* Add integration tests for unauthorised calls
* Ignore .vscode
* Add gitattributes to fix UI eol lint issue
* Misc changes
* Prevent caching of integration test results
* Disallow changing root name or roles
* Remove old api key config
* Ensure api key matches stored key
* Add session management
* Fix empty key detection code
* Allow users to query users, hiding some info
* Add regenerate api key and change password

WithoutPants committed:
Remove sqlite (#14)

Commit:
Omnibus stashdb changes (#19)

* Add fuzzy performer/scene search
* Add frontend support for users
* Add support for tags and fingerprints
* Fingerprint duration and UI updates
* Add image entity for performers/scenes/studios
* Switch frontend to create-react-app and typescript 3.9
* Remove image fields from URL type
* Remove image fields from studio
* Fix bugs, and update frontend config/readme
* Add pg_trgm info
* Fix integration tests
* Create image url resolver and remove unused image fields
* Don't open browser by default
* Fix development auth cookie issues

Co-authored-by: WithoutPants <53250216+WithoutPants@users.noreply.github.com>

Leopere committed:
Correcting FUNDING.yml

There was drift after a decision made to funnel people into Open Collective rather than librepay.

Commit:
Fix bugs and allow for unsetting performer metadata. (#20)

* Fix various editing/display bugs
* Add entity create buttons and fix button style/behaviors
* Show country name in performer info

WithoutPants committed:
Add some docker files and travis config (#22)

Commit:
Add editing support for tags (#21)

* Schema/api changes for edit
* Add StrSliceCompare
* Edit Tag mutation
* Tag edit creation
* ApplyEdit for tags
* Add migration
* Edit comments
* Add edits to tag resolver
* QueryEdit filters
* Add frontend tag editing support
* Upgrade apollo-client to version 3
* Upgrade react-router to 5.2.0
* Upgrade react-hook-form and yup
* Update various libraries
* Search fixes
* Show tag description and aliases
* Update libraries
* Move scene_tags in tag migrations
* Add edits menu item and tweak graphql schema
* Show merge edits for tags that have been merged

Co-authored-by: WithoutPants <53250216+WithoutPants@users.noreply.github.com>

Commit:
Add batch fingerprint query endpoint (#23)

* Add batch fingerprint query endpoint
* Make sure duplicate scenes aren't returned.

WithoutPants committed:
Merge branch 'develop' into stash_prototype

WithoutPants committed:
Merge pull request #24 from stashapp/stash_prototype

Merge stash prototype into develop

WithoutPants committed:
Update api_key

WithoutPants committed:
Second attempt at apikey

WithoutPants committed:
Another attempt at api key

Commit:
Add OSHASH algorithm type (#26)

Commit:
Add dataloaders (#25)

Commit:
Query optimizations (#27)

* Change default scene sort to date DESC

WithoutPants committed:
User registration (#31)

* Add configuration keys
* Move user stuff to separate package
* Add grant/repeal token implementation
* Invite key generation
* Initial new user implementation
* Send activation email
* Add new user activation implementation
* Allow resending activation email
* Return activation key if verification not required
* Implement reset password
* Add generate-ui make target
* Invite token and code management
* Restore original generated ts files
* Don't format generated files
* Add register page
* Add activate new user page
* Add reset password UI
* Add config documentation

Commit:
Support uuids in search terms (#32)

Commit:
Break email port out into separate config field (#36)

WithoutPants committed:
Image local storage and editing (#33)

Commit:
User components cleanup (#38)

Commit:
Add s3 image storage (#39)

* Make width/height/checksum non-nullable
* Add s3 image backend
* Change local to file
* Update golang version in travis from 1.11 => 1.13

Commit:
WIP

Commit:
Push prod/dev images to dockerhub (#40)

Commit:
Rename namespace to stash-box and update linux compiles to be static (#41)

Commit:
Update travis cross compile command (#42)

Commit:
Correct cross-compile compiler version (#43)

Commit:
Update dockerfile filename and command (#44)

Commit:
Add performer editing (#35)

Commit:
Date and duration parsing

Commit:
Add Tag Categories (#28)

Commit:
WIP

Commit:
UI update (#45)

Commit:
Bug fixes (#46)

* Update libraries

* Fix enter form handling so it enter can be used in edit notes

* Add required editnote to all edit pages

* Fix using enter in edit forms

* Preserve whitespace in edit comments

* Fix performer info header wrapping

* Fix unsetting of performer metadata

* Yarn format

Commit:
Fix unsetting of performer metadata (#47)

Commit:
Order scene images for landscape aspect ratio (#48)

Commit:
Add edit commenting and fix various bugs (#49)

* Show performer disambiguation in scene form

* Shrink page header size

* Filter out duplicate aliases in performer merges

* Make scene/studio edit buttons admin only

* Fix crash when erasing search input field

* Disable generate invite button if user has no tokens

* Fix bug where updating user would wipe roles

* Make edit user button admin only

* Add edit commenting

Commit:
Add performer edit options to control performance aliases (#50)

Commit:
Require admin perm for users list, and prune user view (#52)

Commit:
Add support for PHash (#53)

Commit:
Sort performer images by portrait aspect ratio by default (#54)

Commit:
Change imagecreation permissions requirements to edit (#55)

peolic committed:
Fix typo (#56)

peolic committed:
Fix incorrect table reference (#57)

* Fix incorrect table reference

* Add test

Co-authored-by: Infinite <infinitekittens@protonmail.com>

Commit:
Frontend fix roundup (#58)

peolic committed:
generate-dataloaders: fix missing line-continuation (#65)

peolic committed:
Expand edit note on performer form (#64)

Gives a little bit more "room" to write info.

jeremymeyers committed:
Clearing up confusion about who needs to install (#63)

peolic committed:
Don't show performance aliases checkbox on create (#60)

Commit:
Fix roundup (#66)

* Prevent submission of fingerprints with 0s duration

* Escape email address in activation email link

* Sort edit comments by timestamp

* Use local timezone for edit/editnote timestamps

* Order user list ascending

* Disable submit buttons while edits are submitting.

* Fix performer/scene update triggers to concatenate and update correctly

* Add scene duration to scene page and search card

* Allow unsetting studio parent id

* Fix performer ordering by scene_count

* Add performer order by created_at

* Fix updating of column names with reserved names

* Allow users to cancel own edits

* Fix edit diff of partial performer birthdates

* Fix birthdate showing up in edit diffs when being removed

* Add career start ordering for performers

* Support searching by tag name with uuid

* Change edit diff checkboxes to icons for visibility

* Add markdown to edit comments

* Add scene filtering by fingerprint

* Switch markdown library to react-markdown

* Prevent generation of image uuids that start with ad

* Fix timestamp timezone

* Restore #60

Co-authored-by: peolic <66393006+peolic@users.noreply.github.com>

* Fix breasttype issues for male performers

* Fix breasttype issue for performers with null breasttype

* Add shortcut for changing performers in scenes

* Linting

* Coalesce breast type to Unknown rather than NA

* Fix breasttype diff logic

* Fix text truncation of performer name in scene form

Co-authored-by: peolic <66393006+peolic@users.noreply.github.com>

peolic committed:
Add basic sorting functionality to the edits page (#61)

Commit:
Fix performer ordering and add index to scene_search (#68)

jeremymeyers committed:
missed a word (#67)

Commit:
Move CI pipeline to Github actions (#70)

Commit:
Fix github action (#77)

Commit:
Github action typo (#79)

Commit:
Make edit users nullable (#76)

peolic committed:
Fix CORS errors in development mode (#71)

peolic committed:
Update frontend development instructions (#74)

* Update frontend development instructions

* Add "engines" for minimum node & yarn requirements

* Remove windows hash from yarn link

* Update Frontend development section on main README

peolic committed:
Fix issue when using `parentStudio` filter (#73)

Commit:
Make linux build static, remove debug symbols, and bundle assets (#80)

Commit:
Add distance matching for phashes (#81)

Commit:
Add submissions count/time tracking to fingerprints (#72)

peolic committed:
implement version resolver (#83)

peolic committed:
Refactor scene form and add duration and director (#82)

Commit:
Add mergedIDs field to performers to track redirected ids (#85)

Commit:
Upgrade gqlgen to 0.13 (#86)

peolic committed:
Fix missing GQL introspection extension (#87)

Commit:
Add linting and additional error checkers (#84)

peolic committed:
Prevent creating body mod if it already exists (#88)

peolic committed:
Add formatted duration to fingerprints (#90)

Commit:
Update postgres dockerfile to use debian instead of alpine (#89)

Commit:
Add studio filter to performer scene list (#91)

peolic committed:
Update cross build command (#94)

* Update cross build command

`golang-cross` now uses the GitHub Package Registry.
https://github.com/gythialy/golang-cross/commit/f205a9b10332342266099850058c402cd22bf2b1

* Use GHCR instead

peolic committed:
User's edits view (#93)

Commit:
Fix studio name not being updated in search table for scene updates (#96)

WithoutPants committed:
Major refactor (#98)

peolic committed:
don't split scene performer names in the middle (#97)

peolic committed:
Add missing gender icon on performer cards (#101)

peolic committed:
Fix manually edited go.mod/sum (#102)

should be able to run `go mod tidy` safely

peolic committed:
Allow creating a new alias with a different casing (#103)

Commit:
Refactor config (#104)

Commit:
Resolve tag aliases and performer redirects

peolic committed:
Fix CI build (#107)

* format `.goreleaser.yml`

* add `go mod tidy` pre-hook

https://goreleaser.com/customization/build/#go-modules

* actually store go build cache

* use `fetch-depth: 0`

peolic committed:
Prevent duplicate alias when changing scene performer (#108)

peolic committed:
Misc backend changes (#106)

* split out enum functions to utils

* split out ApplyEdit to a function

* fix InsertJoinsWithConflictHandling comment

* getRepoFactory once

Co-authored-by: Infinite <infinitekittens@protonmail.com>

Commit:
WIP

peolic committed:
Fix cached performer studio select (#109)

peolic committed:
Fix PerformerForm birthdate typo (#110)

peolic committed:
Ignore unset `max_dimension` value (#111)

when generating an S3 image URL

peolic committed:
Update s3image.go (#112)

WithoutPants committed:
Graphql schema changes

Commit:
Add name filtering to users list (#117)

* Add name filtering to users list

* Add perPage param to users list

Co-authored-by: peolic <66393006+peolic@users.noreply.github.com>

Co-authored-by: peolic <66393006+peolic@users.noreply.github.com>

Commit:
Replace packr with go:embed (#118)

Commit:
Fix github pipeline (#119)

peolic committed:
Fix changing description on an existing body mod location (#105)

* fix changing description on an existing body mod location

* Fix comparison

Co-authored-by: Infinite <infinitekittens@protonmail.com>

Commit:
Update frontend libraries (#114)

WithoutPants committed:
Graphql changes

WithoutPants committed:
Schema changes

WithoutPants committed:
Add new user roles

WithoutPants committed:
Rewrite import code

WithoutPants committed:
Fix backend bugs

WithoutPants committed:
Update UI

WithoutPants committed:
Backend bug fixes

WithoutPants committed:
Add make targets

WithoutPants committed:
Lint/format

peolic committed:
Fix `yarn generate` "spurious results" error (#121)

WithoutPants committed:
Merge remote-tracking branch 'upstream/develop' into bulk-import-ext

ghost commented:
Thanks for picking this up!

Saving the import data in the db is a good idea. I would expand on that and allow multiple pending imports per user, so it's possible to for instance automatically upload datasets with preset configs for manual tweaking/approval. 

I also think it would help a lot if it was possible to modify the config after an import has been uploaded. Ideally it would just be a tabbed interface with the config in one pane, and the output in the other, so you could quickly modify regular expressions and see the updated results.

Regarding the main studio field that I added, the intention was that it could represent the network in the case where scenes might be from different sub studios of a scraped site. In your model we could accomplish the same thing with an optional parent studio field, then the parent studio could be fixed and the studio field dynamically set per scene.

It's no longer possible to parse durations from what I can tell. I used regexes with named capture groups, so you could do `(?P<hours>\d+)h, = (?P<minutes>\d+) minutes` for instance, to parse a particular duration format a site might use. I don't think replacement fields are sufficient for that case. The same goes for dates, I used the `dateparse` library for that.

ghost commented:
> * importing scenes with an image URL is super slow and probably not done in a sustainable way - it currently just slurps all the URLs as needed. This probably needs to be offloaded to a separate task and throttled appropriately. An alternative is to accept encoded images in the data files and just use those instead of scraping them from the source.

The latter is what I ended up doing when scraping. Throw all images in an async queue and let them finish on their own accord, otherwise it holds up the process too much.

EDIT: The former I mean. Uploading images with the data would work, although it complicates the process a bit.

bnkai commented on pkg/manager/bulkimport/csv.go:
```diff
@@ -0,0 +1,37 @@
+type csvParser struct{}
+
+func (csvParser) parse(input io.Reader, fn processFn) error {
+	reader := csv.NewReader(input)
```
For importing large csv files setting `ReuseRecord` reduces the allocations a lot. It shouldnt cause any trouble since every row seems to be processed and then discarded anyway.
```golang
	reader := csv.NewReader(input)
  csvreader.ReuseRecord = true
```

WithoutPants commented:
> Regarding the main studio field that I added, the intention was that it could represent the network in the case where scenes might be from different sub studios of a scraped site. In your model we could accomplish the same thing with an optional parent studio field, then the parent studio could be fixed and the studio field dynamically set per scene.

We're only setting a single studio on a scene. Do you mean a default studio if no specific studio is set? Adding a default field value shouldn't be too difficult, if I understand correctly.

> I also think it would help a lot if it was possible to modify the config after an import has been uploaded. Ideally it would just be a tabbed interface with the config in one pane, and the output in the other, so you could quickly modify regular expressions and see the updated results.

I'll have to reconsider the upload process. At the moment, it requires the config while uploading so that we can eliminate unnecessary data. Instead, the process will have to be to upload the file first, then select the columns and options, with the ability to preview the results.

peolic committed:
Fix body mods (#122)

ghost commented:
> > Regarding the main studio field that I added, the intention was that it could represent the network in the case where scenes might be from different sub studios of a scraped site. In your model we could accomplish the same thing with an optional parent studio field, then the parent studio could be fixed and the studio field dynamically set per scene.
> 
> We're only setting a single studio on a scene. Do you mean a default studio if no specific studio is set? Adding a default field value shouldn't be too difficult, if I understand correctly.

I mean when creating new studios. For instance when importing a report from studio A, which has sub-studios B,C,D, etc denoted in the column `studios`. In this case you would set the `parent studio` field to A so that if a new studio is encountered it would get the correct parent.

> > I also think it would help a lot if it was possible to modify the config after an import has been uploaded. Ideally it would just be a tabbed interface with the config in one pane, and the output in the other, so you could quickly modify regular expressions and see the updated results.
> 
> I'll have to reconsider the upload process. At the moment, it requires the config while uploading so that we can eliminate unnecessary data. Instead, the process will have to be to upload the file first, then select the columns and options, with the ability to preview the results.

Right, I didn't consider that we have to store the raw report. I think that should be fine though. An advantage of uploading first is that you can sniff the column names and populate title/performers/duration, etc. in the column options immediately.

WithoutPants commented:
> I mean when creating new studios. For instance when importing a report from studio A, which has sub-studios B,C,D, etc denoted in the column `studios`. In this case you would set the `parent studio` field to A so that if a new studio is encountered it would get the correct parent.

Rightio. The current iteration doesn't handle creating studios/performers/tags as part of the scene import process, so I might defer that to a later iteration.

WithoutPants committed:
Update backend

WithoutPants committed:
Refactor import process

WithoutPants committed:
List delimiter per field

WithoutPants committed:
Reuse csv records

WithoutPants committed:
Validate image location during image creation

WithoutPants committed:
Parse dates and durations

WithoutPants committed:
Sort fields

WithoutPants committed:
One txn per scene. Log imports

WithoutPants committed:
Format

peolic committed:
Fix adding comments to existing edits (#123)

* Fix adding comments to existing edits

* Disable comment save button if empty comment

WithoutPants commented:
Changed the process to be three-step: upload file, massage the data, then map and complete the import. 

The massage view shows the field config like in the first screenshot, plus a view of the data rows. Shows matched and unmatched fields:
![image](https://user-images.githubusercontent.com/53250216/130873509-d596e649-86ef-4861-9426-f4519779ee07.png)

Moved list delimiter to be per-field.

Import process is still single-threaded and very slow, but it now logs the scenes as they are imported.

peolic commented on Makefile:
```diff
@@ -88,6 +88,14 @@ pre-ui:
+	cd frontend && yarn format
+
+.PHONY: ui-start
+ui-start: pre-build
```
Did you mean?
```suggestion
ui-start: pre-ui
```

WithoutPants replied:
No. `pre-build` is used to populate the build environment variables. In stash these are passed into the react compile process to embed version information. It is not currently done here, but it doesn't hurt that its there.

Commit:
Fix scene performer changing (#125)

Commit:
Add editing of usernames for admins (#126)

WithoutPants committed:
Studio edits (#127)

* Refactor edit code
* Remove child studio mutations
* Studio edit processing
* Refactor/reduce boilerplate
* Add integration test
* Add studio to edit query
* Make Studio in list mandatory
* Exclude delete studios from query
* Add UI pages
* Add missing build directives
* Fix application of image/url studio edits
* Add diff of studio urls and edit target header link
* Fix validate error message
* Fix destroy edit target link

Co-authored-by: Infinite <infinitekittens@protonmail.com>

WithoutPants committed:
Fix studio redirect table spec

WithoutPants committed:
Merge remote-tracking branch 'upstream/develop' into bulk-import-ext

WithoutPants commented:
I think this really needs a job management system to work well. I'm going to defer this and add job management, so that the import system can use it.
